drivers: gpio: add dts support for nrf52 gpio

marc-cpdesign opened the pull request:
This adds basic support for declaring gpio nodes in dts for nrf52.
The dts.fixup provides mapping for the generated defines to the config
defines currently used by the nrf gpio driver.
A new Kconfig option "HAS_DTS_GPIO" is introduced so that it can be
defined by boards which want to opt-in to using the dts to specify
gpio configuration.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

nvlsianpu commented on drivers/gpio/Kconfig.nrf5:
```diff
@@ -26,7 +26,7 @@ config GPIO_NRF5_P0
 config GPIO_NRF5_P0_DEV_NAME
 	string "GPIO Port P0 Device Name"
-	depends on GPIO_NRF5_P0
+	depends on GPIO_NRF5_P0 && !HAS_DTS_GPIO
```
This is just suggestion: how about grouping all properties dependent from !HAS_DTS_GPIO using `if !HAS_DTS_GPIO ` instead of adding dependency for each other. This will improve legibility of this file.

marc-cpdesign replied:
Yes, I think I agree with you. I took the precedent from some existing Kconfig, which I've just looked again and it was from the I2C and ADC drivers. Looking at them again, I think this would apply to them also.

Will change.

pizi-nordic commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -54,6 +54,25 @@
+		gpio0: gpio@50000000 {
+			compatible = "nordic,nrf5-gpio";
+			gpio-controller;
+			interrupts = <6 5>;
```
Is there a reason to put GPIO interrupts on other priority than everything else?

Also, according to spec the GPIO block cannot issue an interrupt: This is done by GPIOTE block, which is not present in the dtsi file. IMHO we should:
- Add gpiote instance to the dts and place interrupt there.
- Add reference from gpio instance to gpiote instance.

marc-cpdesign replied:
Hi, Thanks for reviewing.

The interrupt priority is exactly what the GPIO driver currently uses as default; this was deliberate to reduce friction in introducing this to dts.

The driver code for the GPIO and GPIOTE are in the same module. Are you asking that just an extra gpiote node should be added to the dts (to be used in existing driver), or that the driver also needs to be split up?

I'm new to zephyr, and the way dts is used is still unfamiliar, but I think that using dts rather than kconfig options is good, and adding dts gpio support is a necessary start for many other drivers that used gpios/pins (pwm,i2c,spi,uart etc) - however as I'm not very knowledgeable yet I'm very much just keeping status quo. Any extra advice is appreciated.

galak reviewed (changes requested):
One minor change to the dts.fixup

marc-cpdesign committed:
drivers: gpio: add dts support for nrf52 gpio

This adds basic support for declaring gpio nodes in dts for nrf52.
The dts.fixup provides mapping for the generated defines to the config
defines currently used by the nrf gpio driver.

Existing boards that use nrf52 are updated.

Signed-off-by: Marc Reilly <marc@cpdesign.com.au>

marc-cpdesign commented on arch/arm/soc/nordic_nrf/nrf52/dts.fixup:
```diff
@@ -7,6 +7,12 @@
 
 #define FLASH_DEV_NAME			NRF_NRF52_FLASH_CONTROLLER_4001E000_LABEL
 
+#ifdef CONFIG_HAS_DTS_GPIO
```
It is still needed at this stage. The existing nrf52 boards are not selecting the HAS_DTS_GPIO option yet, and so they still have defines for the _DEV_NAME coming from Kconfig. This causes the defines in the fixup file to be redefines, and compilation fails. (

I think the options are:
- Leave this here for now, and remove when possible later
- Update all the nrf52 board dts files to enable the gpio node, and remove the _DEV_NAME from Kconfig

(no surprise, I prefer the first option :) )

marc-cpdesign replied:
Done.

carlescufi commented:
@galak @pizi-nordic just saw this, looks good to me. Can we re-review to try and get it in 1.12?

carlescufi commented:
@pizi-nordic can you revisit this?

carlescufi commented:
@marc-cpdesign a couple of outstanding PRs and upcoming work are dependant on this PR. Can you let us know if you plan to update it soon? Otherwise we can take over it, while keeping your commits and signed-off-by.

marc-cpdesign commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -55,6 +55,25 @@
+			compatible = "nordic,nrf5-gpio";
+			gpio-controller;
+			interrupts = <6 5>;
+			reg = <0x50000504 0x27C>;
```
I will try separate the gpiote into a different node.

For the reg property address:
I'm wasn't sure what to do here, as the instance for gpio0 (P0 in doc) is at 0x5000_0000 and the instance for gpio1 (P1 in doc) is at 0x5000_0300, but the gpio registers only start at offset 0x504.

The device tree spec says: 
"The reg property describes the address of the device’s resources within the address space defined by its parent bus." (https://github.com/devicetree-org/devicetree-specification/blob/master/source/devicetree-basics.rst)

So, I think the current way it is is acceptable, although that's much to do with me not knowing how to do it differently without having overlapping regions for gpio0 and gpio1

marc-cpdesign commented:
I've updated with my latest attempt at trying to split out gpio and gpiote in the dts, but it doesn't seem like any defines are being generated for the gpiote node.

I'll keep on at it tomorrow, but if it is holding people up and they can fix it then please feel free to take over!

carlescufi commented:
@marc-cpdesign I took a look at this and @erwango helped me debug, here's a patch:

```
$ git diff
diff --git a/dts/arm/nordic/nrf52832.dtsi b/dts/arm/nordic/nrf52832.dtsi
index e58278c98..24a9a8ba8 100644
--- a/dts/arm/nordic/nrf52832.dtsi
+++ b/dts/arm/nordic/nrf52832.dtsi
@@ -47,7 +47,7 @@
                };

                gpiote0: gpiote@40006000 {
-                       compatible = "nordic,nrf-gpiote";
+                       compatible = "nordic,nrf5-gpiote";
                        reg = <0x40006000 0x1000>;
                        interrupts = <6 5>;
                        status = "disabled";
diff --git a/dts/bindings/gpio/nordic,nrf5-gpiote.yaml b/dts/bindings/gpio/nordic,nrf5-gpiote.yaml
index 94c46e43e..6252f0736 100644
--- a/dts/bindings/gpio/nordic,nrf5-gpiote.yaml
+++ b/dts/bindings/gpio/nordic,nrf5-gpiote.yaml
@@ -5,6 +5,7 @@
 #
 ---
 title: NRF5 GPIOTE
+id: nordic,nrf5-gpiote
 version: 0.1

 description: >
```

erwango commented on dts/arm/nordic/nrf52832.dtsi:
```diff
@@ -46,6 +46,24 @@
 		};
 
+		gpiote0: gpiote@40006000 {
+			compatible = "nordic,nrf-gpiote";
```
nordic,nrf**5**-gpiote

marc-cpdesign replied:
Thanks! I can't believe how many times I looked over that. Quite embarrassing :|

erwango replied:
Frequent issue :-) you're not the only one (count me in!)

erwango commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -55,6 +55,33 @@
+			compatible = "nordic,nrf-gpiote";
+			reg = <0x40006000 0x1000>;
+			status = "disabled";
+			label = "GPIOTE_0";
```
What could be interesting here is to add a channel and bind it to a gpio pin.

marc-cpdesign replied:
I left this for a moment, with the idea of expanding the capabilities later as they are needed

erwango commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -55,6 +55,33 @@
+		gpio0: gpio@50000000 {
+			compatible = "nordic,nrf5-gpio";
+			gpio-controller;
+			gpiote = &gpiote0;
```
What relationship are you trying to enable here?

marc-cpdesign replied:
I added it initially due to earlier comments from pizi-nordic, and it seemed like a good idea to express some relationship between the gpio and gpiote. 
However its a bit redundant, so I took it out.

erwango replied:
Is that configurable? I mean can we configure gpio0 to work with gpiote1?

erwango commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -55,6 +55,31 @@
+			reg = <0x40006000 0x1000>;
+			status = "disabled";
+			label = "GPIOTE_0";
+		};
```
If gpiote channel configuration is done in a gpiote driver along with reference to a specific gpio pin,
I'd expose that in the following way, as a subnode of gpiote
```
                gpio0pin1event0: channel@510 {
                          reg = <0x510>;
                          gpio = <&gpio0 1>;
                          mode = "event";
                          polarity = "HiToLo";
                }
```
Though, if we consider each gpiote if associated to a gpio (since IIUC we cannot configure GPIO port in GPIOTE register), gpiote0 can maybe be set as a subnode of gpio0:
```
		gpio0: gpio@50000000 {
			compatible = "nordic,nrf5-gpio";
			gpio-controller;
			reg = <0x50000000 0x1000>;
			#gpio-cells = <2>;
			label = "GPIO_0";
			status = "disabled";

		         gpiote0: gpiote@40006000 {
			         compatible = "nordic,nrf5-gpiote";
			         reg = <0x40006000 0x1000>;
			         status = "disabled";
			         label = "GPIOTE_0";

                                gpio0pin1event0: channel@510 {
                                            reg = <0x510>;
                                            mode = "event";
                                            polarity = "HiToLo";
                               };

         		};
		};
```

carlescufi replied:
@pizi-nordic, @anangl, @kl-cruz  mind to comment on the above?

carlescufi replied:
@erwango From what I can tell GPIO and GPIOTE are completely separate, independent peripherals. When there's a single port (nRF52832 for example) then one selects the pin with `PSEL` in GPIOTE in order to configure an interrupt or task/event connection. If there's more than one port (nRF52840), one has to configure both the port and the pin in the GPIOTE config.
http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52832.ps.v1.1/gpiote.html?cp=2_1_0_20_3_2#register.CONFIG-0
http://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52840.ps/gpiote.html?cp=2_0_0_5_9_3_2#register.CONFIG-0-7

erwango replied:
Ok, I had seen only the on port specification.
Then I think following would make sense as an individual node (not gpio subnode)
```
		         gpiote0: gpiote@40006000 {
			         compatible = "nordic,nrf5-gpiote";
			         reg = <0x40006000 0x1000>;
			         status = "disabled";
			         label = "GPIOTE_0";

                                gpio0pin1event0: channel@510 {
                                            reg = <0x510>;
                                            gpios = <&gpio0 1>
                                            mode = "event";
                                            polarity = "HiToLo";
                               };
```

One thing to check is the gpio-cells dimension (here I assumed gpio-cells = 1), as generally this is 2 but last slot is used for IRQ (which is not available here if IUC). But it could be replaced with any other information pertinent to gpio way of working.

marc-cpdesign replied:
At the moment the gpio-cells is 2, for pin and flags.

```
                            gpiote0: gpiote@40006000 {
			        compatible = "nordic,nrf5-gpiote";
			        reg = <0x40006000 0x1000>;
			        status = "disabled";
			        label = "GPIOTE_0";

                                gpiote_chan0: channel@510 {
                                            reg = <0x510>;
                                            status = "disabled";
                                };

                                gpiote_chan1: channel@514 {
                                            reg = <0x514>;
                                            status = "disabled";
                                };
```

.. and then in the board:
```
    &gpiote_chan0 {
        gpios = <&gpio0 1 0>
        mode = "event";
        polarity = "HiToLo";
        status = "ok";
    };
```

That said, I'm not sure how this will all work to enable other drivers (for example an I2C sensor) to specify a gpio for interrupt source.

erwango replied:
> That said, I'm not sure how this will all work to enable other drivers (for example an I2C sensor) to specify a gpio for interrupt source.

You should add, in gptiote:
```
			interrupt-controller;
			#interrupt-cells = <x>;
```
In the node that need an irq:
```
			interrupt-parent = <&gpiote_chan0>;
			interrupts = <1>;
			interrupt-names = "bla";
```
Note that I am unsure how the interrupts should be populated when you have only one line.
Maybe you can remove the channels subnode and do, in i2c slave node, something like:
```
			interrupt-parent = <&gpiote>;
			interrupts = <1 2 3>;
			interrupt-names = "bla";
```
with, intterupt-cells = <3> in gpiote that could hold the following fields
1: channel
2: event
3: polarity

marc-cpdesign commented:
recheck

carlescufi commented:
recheck

marc-cpdesign commented:
@pizi-nordic I've updated the nodes as you requested
@erwango I experimented with adding channels to the gpiote, but nothing came together easily.

pizi-nordic commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -55,6 +55,32 @@
+		gpio0: gpio@50000000 {
+			compatible = "nordic,nrf5-gpio";
+			gpio-controller;
+			reg = <0x50000000 0x1000>;
```
The gpio0 memory region (reg property) overlaps with the gpio1 memory region.

marc-cpdesign replied:
Hi, thanks for reviewing.

Like I said earlier, the regions for the two must overlap if we want the reg address to be the base address of the IP block, because the base address of gpio1 is in the gpio0 range:

0x5000_0000 --> gpio0 base address
0x5000_0300 --> gpio1 base address
0x5000_0504 --> first gpio0 register
0x5000_0804 --> first gpio1 register

This is why in the first go around I had the addresses as 0x50000504 and 0x50000804, although this doesn't make much sense when looking at it from the generated defines/driver perspective.

Options:
- ignore the problem
- make the size of the reg 0, (or leave the size out?)
- go back to the way I first had it
- make gpio0 and gpio1 children of a new node type:
```
gpio {
    compatible="nordic,nrf5-gpio-????";
    label = "????";
    reg = <0x50000000 0x1000>;

    gpio0 {
       compatible="nordic-nrf5-gpio";
       offset = <0x504>;  
       gpio-controller;
       label = "GPIO0";
    };
    gpio1 {
       compatible="nordic-nrf5-gpio";
       offset = <0x804>;
       gpio-controller;
       label = "GPIO1"
    };
};
```

pizi-nordic replied:
@marc-cpdesign: Thank you for clear explanation. Now I see the problem.  In such situation we should declare several regions in the reg property:

```
gpio0: gpio@50000000 {
        reg = <0x50000000 0x300
                    0x50000500 0x300>;
}
```

However I do not know if such declarations are supported by the Zephyr DTS infrastructure.
@nvlsianpu, @carlescuf: Do you know if our infrastructure supports entries like the one above?

nvlsianpu replied:
This is what I've got:
```
/* gpio@50000000 */
#define NORDIC_NRF5_GPIO_50000000_BASE_ADDRESS_0	0x50000000
#define NORDIC_NRF5_GPIO_50000000_BASE_ADDRESS_1	0x50000500
#define NORDIC_NRF5_GPIO_50000000_LABEL			"GPIO_0"
#define NORDIC_NRF5_GPIO_50000000_SIZE_0		768
#define NORDIC_NRF5_GPIO_50000000_SIZE_1		768
```
@pizi-nordic  so It is supported.

marc-cpdesign replied:
@nvlsianpu Thanks for that. I will make changes shortly.

marc-cpdesign commented:
@pizi-nordic Register changes are done.
@erwango Are you able to have another look at this please.

erwango commented on arch/arm/soc/nordic_nrf/nrf52/dts.fixup:
```diff
@@ -7,6 +7,10 @@
 
+#define CONFIG_GPIO_NRF5_P0_DEV_NAME	NORDIC_NRF5_GPIO_50000000_LABEL
+#define CONFIG_GPIO_NRF5_P1_DEV_NAME	NORDIC_NRF5_GPIO_50000300_LABEL
+#define CONFIG_GPIOTE_NRF5_PRI		NORDIC_NRF5_GPIOTE_40006000_IRQ_0_PRIORITY
```
Since you're moving gpio to dts, I think you should get IRQ number too from dts (and remove the one from CMSIS)

marc-cpdesign replied:
What do you mean by this? I don't see what needs to be converted in the dts.fixup

do you mean in gpio_nrf5.c, that NRF5_IRQ_GPIOTE_IRQn should be changed to NORDIC_NRF5_GPIOTE_40006000_IRQ_GPIOTE ?
```
static int gpio_nrf5_init(struct device *dev)
{
	IRQ_CONNECT(NRF5_IRQ_GPIOTE_IRQn, CONFIG_GPIOTE_NRF5_PRI,
		    gpio_nrf5_port_isr, NULL, 0);

	irq_enable(NRF5_IRQ_GPIOTE_IRQn);

	return 0;
}
```

erwango replied:
Indeed, but you'll need to go through a KConfig symbol, like CONFIG_GPIOTE_NRF5_IRQ

marc-cpdesign replied:
I don't see the point

erwango replied:
Aim is to extract driver configuration from device tree, and use it as a reference.
Having information coming from different sources (dts and CMSIS) will be confusing at the end.
Such as defining it in device tree and not using it.
Today, there is information coming from different sources but aim is to converge to device tree.

Mierunski replied:
There should be 2 defines regarding IRQ in fixup
```
#define CONFIG_GPIOTE_NRF5_IRQ		NORDIC_NRF5_GPIOTE_40006000_IRQ_0
#define CONFIG_GPIOTE_NRF5_IRQ_PRI	NORDIC_NRF5_GPIOTE_40006000_IRQ_0_PRIORITY
```
And then change references to IRQ in gpio_nrf5.c
```
static int gpio_nrf5_init(struct device *dev)
{
	IRQ_CONNECT(CONFIG_GPIOTE_NRF5_IRQ, CONFIG_GPIOTE_NRF5_IRQ_PRI,
		    gpio_nrf5_port_isr, NULL, 0);

	irq_enable(CONFIG_GPIOTE_NRF5_IRQ_PRI);

	return 0;
}
```
This way it will be consistent with dts for other peripherals

Mierunski replied:
LGTM now 
@erwango could you please revisit?

erwango reviewed (dismissed):
One comment otherwise sounds good.
